services/src/main/java/org/keycloak/authorization/authorization/AuthorizationTokenService.java

```diff
@@ -174,29 +174,47 @@ public static AuthorizationTokenService instance() {
     }
 
     private static void fireErrorEvent(EventBuilder event, String error, Exception cause) {
-        event.detail(Details.REASON, cause == null || cause.getMessage() == null ? "<unknown>" : cause.getMessage())
-                .error(error);
+        if (cause instanceof CorsErrorResponseException) {
+            // cast the exception to populate the event with a more descriptive reason
+            CorsErrorResponseException originalCause = (CorsErrorResponseException) cause;
+            event.detail(Details.REASON, originalCause.getErrorDescription() == null ? "<unknown>" : originalCause.getErrorDescription())
+                    .error(error);
+        } else {
+            event.detail(Details.REASON, cause == null || cause.getMessage() == null ? "<unknown>" : cause.getMessage())
+                    .error(error);
+        }
+
         logger.debug(event.getEvent().getType(), cause);
     }
-    
-    public Response authorize(KeycloakAuthorizationRequest request) {
+
+    public Response authorize(KeycloakAuthorizationRequest request, EventBuilder event) {
         if (request == null) {
-            throw new CorsErrorResponseException(request.getCors(), OAuthErrorException.INVALID_GRANT, "Invalid authorization request.", Status.BAD_REQUEST);
+            CorsErrorResponseException invalidAuthorizationRequestException = new CorsErrorResponseException(null, OAuthErrorException.INVALID_GRANT, "Invalid authorization request.", Status.BAD_REQUEST);
+            fireErrorEvent(event, Errors.INVALID_AUTHORIZATION_REQUEST, invalidAuthorizationRequestException);
+            throw invalidAuthorizationRequestException;
         }
 
         // it is not secure to allow public clients to push arbitrary claims because message can be tampered
         if (isPublicClientRequestingEntitlementWithClaims(request)) {
-            throw new CorsErrorResponseException(request.getCors(), OAuthErrorException.INVALID_GRANT, "Public clients are not allowed to send claims", Status.FORBIDDEN);
+            CorsErrorResponseException forbiddenClientException = new CorsErrorResponseException(request.getCors(), OAuthErrorException.INVALID_GRANT, "Public clients are not allowed to send claims", Status.FORBIDDEN);
+            fireErrorEvent(event, Errors.PUBLIC_CLIENT_SENDING_CLAIMS, forbiddenClientException);
```

Comment:
I would just use `Errors.INVALID_REQUEST`. The details should be enough to understand what happened.

Reply:
agreed